in_toto/keylib.go

```diff
@@ -2,220 +2,402 @@ package in_toto
+		// implementation
+		if len(privateKeyBytes) > 0 {
+			k.KeyVal = KeyVal{
+				Private: strings.TrimSpace(string(privateKeyBytes)),
+				Public:  strings.TrimSpace(string(GeneratePublicPemBlock(pubKeyBytes))),
+			}
+		} else {
+			k.KeyVal = KeyVal{
+				Public: strings.TrimSpace(string(pubKeyBytes)),
+			}
+		}
+	case "ed25519":
+		if len(privateKeyBytes) > 0 {
+			k.KeyVal = KeyVal{
+				Private: strings.TrimSpace(hex.EncodeToString(privateKeyBytes)),
+				Public:  strings.TrimSpace(hex.EncodeToString(pubKeyBytes)),
+			}
+		} else {
+			k.KeyVal = KeyVal{
+				Public: strings.TrimSpace(hex.EncodeToString(pubKeyBytes)),
+			}
+		}
+	default:
+		return fmt.Errorf("%w: %s", ErrUnsupportedKeyType, keyType)
+	}
+	k.KeyType = keyType
+	k.Scheme = scheme
+	k.KeyIdHashAlgorithms = keyIdHashAlgorithms
+	if err := k.GenerateKeyId(); err != nil {
 		return err
 	}
+	return nil
+}
 
-	hashed := sha256.Sum256(data)
+/*
+ParseKey tries to parse a PEM []byte slice.
+Supported are:
 
-	// Create hex bytes from the signature hex string
-	sigHex, _ := hex.DecodeString(sig.Sig)
+	* PKCS8
+	* PKCS1
+	* PKIX
 
-	// SecSysLib uses a SaltLength of `hashes.SHA256().digest_size`, i.e. 32
-	if err := rsa.VerifyPSS(rsaPub, crypto.SHA256, hashed[:], sigHex,
-		&rsa.PSSOptions{SaltLength: sha256.Size, Hash: crypto.SHA256}); err != nil {
-		return err
+On success it returns the parsed key and nil.
+On failure it returns nil and the error ErrFailedPEMParsing
+*/
+func ParseKey(data []byte) (interface{}, error) {
+	key, err := x509.ParsePKCS8PrivateKey(data)
+	if err == nil {
+		return key, nil
 	}
-
-	return nil
+	key, err = x509.ParsePKCS1PrivateKey(data)
+	if err == nil {
+		return key, nil
+	}
+	key, err = x509.ParsePKIXPublicKey(data)
+	if err == nil {
+		return key, nil
+	}
+	return nil, ErrFailedPEMParsing
 }
 
 /*
-ParseEd25519FromPrivateJSON parses an ed25519 private key from the json string.
-These ed25519 keys have the format as generated using in-toto-keygen:
+LoadKey loads the key file at specified file path into the key object.
+It automatically derives the PEM type and the key type.
+Right now the following PEM types are supported:
 
-	{
-		"keytype: "ed25519",
-		"scheme": "ed25519",
-		"keyid": ...
-		"keyid_hash_algorithms": [...]
-		"keyval": {
-			"public": "..." # 32 bytes
-			"private": "..." # 32 bytes
-		}
-	}
-*/
-func ParseEd25519FromPrivateJSON(JSONString string) (Key, error) {
+	* PKCS1 for private keys
+	* PKCS8	for private keys
+	* PKIX for public keys
+
+The following key types are supported:
+
+	* ed25519
+	* RSA
+	* ecdsa
 
-	var keyObj Key
-	err := json.Unmarshal([]uint8(JSONString), &keyObj)
+On success it will return nil. The following errors can happen:
+
+	* path not found or not readable
+	* no PEM block in the loaded file
+	* no valid PKCS8/PKCS1 private key or PKIX public key
+	* errors while marshalling
+	* unsupported key types
+*/
+func (k *Key) LoadKey(path string, scheme string, keyIdHashAlgorithms []string) error {
```

Comment:
Can we check if we support the passed keyIdHashAlgorithms and scheme inputs?

Scheme we could actually check twice, once her if it is any supported scheme, and once below after the key type has been assessed, to check if the scheme is supported for the given type.

We should use module global constants for these values.

Reply:
Right now it looks like we only support one scheme per key type.

Reply:
I have added scheme and key type checking.. right now I am only calling this while setting the key components (only on loading.. we may want to call this while signing or validating, too.. right?).